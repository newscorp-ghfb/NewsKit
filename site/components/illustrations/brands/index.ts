import Barrons from './barrons';
import MansionGlobal from './mansion-global';
import Marketwatch from './marketwatch';
import Sun from './sun';
import TalkRadio from './talk-radio';
import TalkSport from './talk-sport';
import TalkTV from './talk-tv';
import Times from './times';
import TimesTravel from './times-travel';
import Virgin from './virgin';
import WSJ from './wsj';

export {
  Barrons,
  MansionGlobal,
  Marketwatch,
  Sun,
  TalkRadio,
  TalkSport,
  TalkTV,
  Times,
  TimesTravel,
  Virgin,
  WSJ,
};
