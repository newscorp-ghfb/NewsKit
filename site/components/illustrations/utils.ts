import {SVGAttributes} from 'react';
import {Theme} from '../../../src/theme';

// @ts-ignore
export const toCSSVar = (theme: Theme, props) => {
  const inlineStyle = {} as SVGAttributes<SVGPathElement>;
  const newProps = {...props};

  if (theme.colors[props?.fill!]) {
    inlineStyle.fill = `var(--color-${props.fill})`;
    delete newProps.fill;
  }
  if (theme.colors[props?.stroke!]) {
    inlineStyle.stroke = `var(--color-${props.fill})`;
    delete newProps.stroke;
  }
  return {
    style: inlineStyle,
    ...newProps,
  };
};

const escapeRegExp = (string: string): string =>
  // $& means the whole matched string
  string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

export const pathToID = (path: string) => {
  if (path) {
    return path
      .replace(new RegExp(escapeRegExp('../'), 'g'), '')
      .replace(new RegExp(escapeRegExp('./'), 'g'), '')
      .replace(new RegExp(escapeRegExp('/'), 'g'), '-');
  }
  return '';
};

export const isSafari =
  typeof window !== 'undefined' &&
  /^((?!chrome|android).)*safari/i.test(window.navigator.userAgent);
