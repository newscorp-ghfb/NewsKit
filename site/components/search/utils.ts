import {DocSearchHit} from '@docsearch/react/dist/esm/types';

export const ignoreFilter = (item: DocSearchHit): boolean => {
  // this includes components overview from results
  if (item.url.includes('components/overview')) {
    return false;
  }

  /*
    Remove introduction result 
    because these 2 results are the same:
    .../components/radio-button/
    .../components/radio-button/#introduction
  */
  if (item.url.endsWith('#introduction')) {
    return false;
  }

  // other filters in the future

  return true;
};

export const mapPathname = (item: DocSearchHit): DocSearchHitExtended => {
  const parseUrl = document.createElement('a');
  parseUrl.href = item.url;

  const canonicalAs = parseUrl.pathname + parseUrl.hash;
  const canonicalPathname = canonicalAs
    .replace(/^\/api/, '/api-docs')
    .replace(/#(.*)$/, '')
    .replace(/\/$/, '');

  return {
    ...item,
    as: canonicalAs,
    pathname: canonicalPathname,
  };
};
