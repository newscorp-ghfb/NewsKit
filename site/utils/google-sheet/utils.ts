import {CMSData, CMSProps, CMSResponse} from './types';

export class CMSError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = 'CMSError';
  }
}

export const formatSheetData = <T extends CMSProps>(data: CMSData): T => {
  const transformedArray = data.map(([key, value]) => [
    key.toLowerCase(),
    value,
  ]);
  // Hack needed to avoid JSON-Serialization validation error from Next.js https://github.com/zeit/next.js/discussions/11209
  // >>> Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.
  return JSON.parse(JSON.stringify(Object.fromEntries(transformedArray)));
};

export const parseCMSResponse = <T extends CMSProps>(
  cmsResponse: CMSResponse,
  {
    required,
    optional = {},
    dynamic = {},
  }: {
    required: Record<string, string>;
    optional?: Record<string, string>;
    dynamic?: Record<string, string>;
  },
): T => {
  if (!cmsResponse) {
    throw new CMSError('No CMS data found');
  }
  const content = formatSheetData<T>(cmsResponse);

  const missingKeys = Object.keys(required).filter(key => !content[key]);
  const invalidKeys = Object.keys(content).filter(key => {
    if (required[key] || optional[key]) {
      return false;
    }
    const prefixMatch = Object.keys(dynamic).find(k =>
      key.match(`${k}\\d+\\b`),
    );
    return !prefixMatch;
  });

  let err = ``;
  if (missingKeys.length) {
    err = `MISSING_KEYS: ${missingKeys.join(', ')}. `;
  }
  if (invalidKeys.length) {
    err = `${err}INVALID_KEYS: ${invalidKeys.join(', ')}. `;
  }

  if (err.length) {
    throw new CMSError(err);
  }

  return content;
};

export const getCMSPropsWithPrefix = <T extends Record<string, string>>(
  content: CMSProps,
  prefix: keyof T & string,
) => Object.entries(content).filter(([k]) => k.startsWith(prefix));
