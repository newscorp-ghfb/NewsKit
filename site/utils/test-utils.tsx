/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme: any = renderToFragmentWithThemeFactory(
  docsThemeLight,
);

export const mockNewsKitComponents = (...components: string[]) => () => ({
  ...(jest.requireActual('newskit') as any),
  ...components.reduce((acc, comp) => {
    acc[comp] = ({children, ...props}: any) => (
      <div
        data-comp={`NewsKit ${comp}`}
        data-testid={comp}
        data-props={JSON.stringify(props, null, 2)}
      >
        {children}
      </div>
    );
    return acc;
  }, {} as Record<string, Function>),
});
