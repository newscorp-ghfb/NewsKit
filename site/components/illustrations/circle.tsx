import {useTheme} from 'newskit';
import React, {SVGAttributes} from 'react';
import {toCSSVar} from './utils';

export const Circle: React.FC<SVGAttributes<SVGCircleElement>> = props => {
  const theme = useTheme();
  return <circle {...toCSSVar(theme, props)} />;
};
