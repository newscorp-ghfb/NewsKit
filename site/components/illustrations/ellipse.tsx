import {useTheme} from 'newskit';
import React, {SVGAttributes} from 'react';
import {toCSSVar} from './utils';

export const Ellipse: React.FC<SVGAttributes<SVGEllipseElement>> = props => {
  const theme = useTheme();
  return <ellipse {...toCSSVar(theme, props)} />;
};
