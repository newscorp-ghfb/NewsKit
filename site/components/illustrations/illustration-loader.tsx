import React, {useEffect, useState} from 'react';
import {isNotSafari, isSafari, pathToID} from './utils';

/**
 * Only for Safari. It reads the svg file as raw string and returns it as an inline SVG.
 */
const getSafariSVG = (
  sanitize: (source: string) => string,
  path: string,
  props?: React.SVGProps<SVGSVGElement>,
) => {
  const rawSVG = sanitize(
    // eslint-disable-next-line import/no-dynamic-require, global-require
    require(`../../public/static/illustrations/${path}.svg`).default,
  );

  return (
    <svg
      viewBox="0 0 1490 838"
      width="100%"
      {...props}
      // eslint-disable-next-line react/no-danger
      dangerouslySetInnerHTML={{__html: rawSVG}}
    />
  );
};

export const getIllustrationComponent = (
  path: string,
  props?: React.SVGProps<SVGSVGElement>,
) => {
  const id = pathToID(path);

  const Component = () => {
    const [safariSVG, setSafariSVG] = useState<JSX.Element | null>(null);

    useEffect(() => {
      if (isSafari && safariSVG === null) {
        // Dynamically importing 'isomorphic-dompurify' only for Safari to avoid bloating the bundle size for other browsers.
        const importSanitizer = async () => {
          const dompurify = await (await import('isomorphic-dompurify'))
            .default;
          const svg = getSafariSVG(dompurify.sanitize, path, props);
          setSafariSVG(svg);
        };
        importSanitizer();
      }
    }, [safariSVG]);

    // If the browser is Safari we render an inline svg.
    const renderSafariSVG = isSafari ? safariSVG : null;

    // If the browser is not Safari we will render an load an external SVG using <use> tag.
    const nonSafariSVG = (
      <svg
        viewBox="0 0 1490 838"
        width={isNotSafari ? '100%' : '0%'}
        {...props}
      >
        <use href={`static/illustrations/${path}.svg#${id}`} />
      </svg>
    );

    return safariSVG === null ? nonSafariSVG : renderSafariSVG;
  };

  return Component;
};

type IllustrationProps = {
  path: string;
} & React.SVGProps<SVGSVGElement>;

export const Illustration: React.FC<IllustrationProps> = ({path, ...props}) => {
  const Component = getIllustrationComponent(path, props);
  return <Component />;
};
