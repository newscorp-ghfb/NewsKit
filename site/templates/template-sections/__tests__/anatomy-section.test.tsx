import React from 'react';
import {getIllustrationComponent} from '../../../components/illustrations/illustration-loader';
import {AnatomySection, AnatomySectionProps} from '..';
import {renderToFragmentWithTheme} from '../../../utils/test-utils';

jest.mock('../../../components/illustrations/illustration-loader', () => ({
  getIllustrationComponent: jest.fn().mockReturnValue(() => <div>Mock</div>),
}));

jest.mock('../../../components/table/table');
jest.mock(
  'newskit',
  require('../../../utils/test-utils').mockNewsKitComponents('Grid', 'Cell'),
);

describe('AnatomySection', () => {
  test('renders section as expected', () => {
    const props: AnatomySectionProps = {
      introduction: 'anatomy introduction',
      media: getIllustrationComponent('https://example.com/image.png'),
      rows: [
        {
          name: 'row 1',
          description: 'description 1',
          component: 'Component1',
          optional: true,
        },
        {
          name: 'row 2',
          description: <span>description 2</span>,
          component: 'Component2',
        },
      ],
    };
    const fragment = renderToFragmentWithTheme(AnatomySection, props);
    expect(fragment).toMatchSnapshot();
  });
});
