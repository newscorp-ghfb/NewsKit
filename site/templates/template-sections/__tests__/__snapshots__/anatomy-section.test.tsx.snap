// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnatomySection renders section as expected 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin-top: -24px;
  padding-top: 24px;
}

@media screen and (max-width: 767px) {
  .emotion-1 {
    margin-bottom: 24px;
  }
}

@media screen and (min-width: 768px) {
  .emotion-1 {
    margin-bottom: 32px;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding: 0;
  background-color: transparent;
}

.emotion-3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media screen and (max-width: 767px) {
  .emotion-4 {
    font-family: "Poppins",sans-serif;
    font-size: 24px;
    line-height: 27px;
    font-weight: 700;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-4::before {
    content: '';
    margin-bottom: -0.2155em;
    display: block;
  }

  .emotion-4::after {
    content: '';
    margin-top: -0.2125em;
    display: block;
  }
}

@media screen and (min-width: 768px) {
  .emotion-4 {
    font-family: "Poppins",sans-serif;
    font-size: 28px;
    line-height: 31.5px;
    font-weight: 700;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-4::before {
    content: '';
    margin-bottom: -0.2155em;
    display: block;
  }

  .emotion-4::after {
    content: '';
    margin-top: -0.2125em;
    display: block;
  }
}

.emotion-5 {
  display: inline;
  font: inherit;
  margin: 0;
  color: #0A0A0A;
}

.emotion-5 svg {
  fill: #0A0A0A;
}

@media screen and (max-width: 1023px) {
  .emotion-6 {
    margin-bottom: 48px;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-6 {
    margin-bottom: 80px;
  }
}

.emotion-7 {
  margin: 0;
  color: #2E2E2E;
  font-family: "Poppins",sans-serif;
  font-size: 20px;
  line-height: 30px;
  font-weight: 500;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-7 svg {
  fill: #2E2E2E;
}

.emotion-7::before {
  content: '';
  margin-bottom: -0.403em;
  display: block;
}

.emotion-7::after {
  content: '';
  margin-top: -0.4em;
  display: block;
}

.emotion-8 {
  margin-bottom: 24px;
}

.emotion-9 {
  width: 100%;
  height: auto;
}

.emotion-10 {
  margin-bottom: 0;
}

<div
    data-comp="NewsKit Cell"
    data-props="{
  \\"xs\\": 12
}"
    data-testid="Cell"
  >
    <section
      class="emotion-0"
      data-toc-indexed="Anatomy"
      id="anatomy"
    >
      <div
        data-comp="NewsKit Grid"
        data-props="{
  \\"xsMargin\\": \\"space000\\",
  \\"xsRowGutter\\": \\"space000\\"
}"
        data-testid="Grid"
      >
        <div
          data-comp="NewsKit Cell"
          data-props="{
  \\"xs\\": 12,
  \\"md\\": 8,
  \\"lg\\": 10,
  \\"xl\\": 8,
  \\"mdOffset\\": 1
}"
          data-testid="Cell"
        >
          <div
            class="emotion-1"
          >
            <div
              class="emotion-2"
            >
              <div
                class="emotion-3"
              >
                <section
                  class="emotion-4"
                >
                  <h2
                    class="nk-headline-heading emotion-5"
                  >
                    Anatomy
                  </h2>
                </section>
              </div>
            </div>
          </div>
          <div
            class="emotion-6"
          >
            <p
              class="emotion-7"
            >
              anatomy introduction
            </p>
          </div>
        </div>
        <div
          data-comp="NewsKit Cell"
          data-props="{
  \\"xs\\": 12,
  \\"mdOffset\\": 1,
  \\"md\\": 10,
  \\"xl\\": 8
}"
          data-testid="Cell"
        />
        <div
          data-comp="NewsKit Cell"
          data-props="{
  \\"xs\\": 12,
  \\"md\\": 10,
  \\"xl\\": 8,
  \\"mdOffset\\": 1
}"
          data-testid="Cell"
        >
          <div
            class="emotion-8"
          >
            <svg
              class="emotion-9"
              fill="none"
              height="1600"
              viewBox="0 0 1600 1600"
              width="1600"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="illustration-Background020"
                height="1600"
                width="1600"
              />
              <path
                clip-rule="evenodd"
                d="M1020.59 586.74H579.412C571.29 586.74 564.706 593.339 564.706 601.48V999.458C564.706 1007.6 571.29 1014.2 579.412 1014.2H1020.59C1028.71 1014.2 1035.29 1007.6 1035.29 999.458V601.48C1035.29 593.339 1028.71 586.74 1020.59 586.74ZM579.412 572C563.168 572 550 585.199 550 601.48V999.458C550 1015.74 563.168 1028.94 579.412 1028.94H1020.59C1036.83 1028.94 1050 1015.74 1050 999.458V601.48C1050 585.199 1036.83 572 1020.59 572H579.412ZM722.249 702.981C722.249 728.415 701.898 749.033 676.794 749.033C651.69 749.033 631.339 728.415 631.339 702.981C631.339 677.547 651.69 656.928 676.794 656.928C701.898 656.928 722.249 677.547 722.249 702.981ZM709.872 832.119C703.597 823.903 691.078 824.573 685.706 833.411L609.592 958.639C603.615 968.472 610.654 981.091 622.116 981.091H979.082C990.092 981.091 997.182 969.356 992.108 959.532L892.423 766.558C887.131 756.315 872.718 755.87 866.812 765.768L786.838 899.781C781.512 908.707 768.916 909.432 762.611 901.176L709.872 832.119Z"
                fill="#E4E4E4"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <div
            class="emotion-10"
          >
            <div>
              Table Component with props:{
  "columns": [
    "Item",
    "Name",
    "Description",
    "Component",
    "Optional"
  ],
  "rows": [
    {
      "name": "row 1",
      "description": "description 1",
      "component": "Component1",
      "optional": true
    },
    {
      "name": "row 2",
      "description": {
        "type": "span",
        "key": null,
        "ref": null,
        "props": {
          "children": "description 2"
        },
        "_owner": null,
        "_store": {}
      },
      "component": "Component2"
    }
  ]
}
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
  @media screen and (max-width: 767px) {
  .emotion-0 {
    margin-top: 64px;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    margin-top: 80px;
  }
}

@media screen and (max-width: 767px) {
  .emotion-0 {
    margin-bottom: 40px;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    margin-bottom: 80px;
  }
}

.emotion-1 {
  border-style: solid;
  border-color: #E4E4E4;
  border-width: 1px;
  border-width: 0px;
  margin: 0;
  border-top-width: 1px;
  width: 100%;
}

<div
    data-comp="NewsKit Cell"
    data-props="{
  \\"xs\\": 12,
  \\"md\\": 10,
  \\"xl\\": 8,
  \\"mdOffset\\": 1
}"
    data-testid="Cell"
  >
    <div
      class="emotion-0"
    >
      <hr
        aria-hidden="true"
        class="emotion-1"
        data-testid="divider"
      />
    </div>
  </div>
</DocumentFragment>
`;
